framework/decode/json_writer.h

```diff
@@ -0,0 +1,155 @@
+#include "util/json_util.h"
+#include "util/platform.h"
+#include "util/defines.h"
+#include "format/format_json.h"
+
+#include "nlohmann/json.hpp"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(util)
+class OutputStream;
+GFXRECON_END_NAMESPACE(util)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+struct ApiCallInfo;
+
+/// Manages writing
+class JsonWriter : public AnnotationHandler
+{
+  public:
+    JsonWriter(const util::JsonOptions& options,
+               const std::string_view   gfxrVersion,
+               const std::string_view   inputFilepath);
+    ~JsonWriter();
+
+    /// Output any data associated with the start of a logical stream such as a header object.
+    void StartStream(util::OutputStream* os);
+    /// Output data at end of stream such as closing the JSON array.
+    void EndStream();
+    void Destroy();
+    bool IsValid() const;
+
+    /// Get a clean fresh JSON tree to build a block's representation in.
+    /// Users can build the JSON representation of their blocks in this tree before
+    /// calling WriteBlockEnd() to serialize it out to the stream.
+    nlohmann::ordered_json& WriteBlockStart();
+
+    /// Finalise the current block and stream it out.
+    void WriteBlockEnd();
+
+    /// Start the JSON tree for a function call, building the top-level object
+    /// with index and function fields, adding name and thread to the function.
+    /// @return The "function" object field for the caller to populate further
+    /// with return value if any and arguments.
+    nlohmann::ordered_json& WriteApiCallStart(const ApiCallInfo& call_info, const std::string_view command_name);
+
+    /// Start the JSON tree for a method call, building the top-level object
+    /// with index and function fields, adding name and thread to the function.
+    /// @return The "method" object field for the caller to populate further
+    /// with return value if any and arguments.
+    nlohmann::ordered_json& WriteApiCallStart(const ApiCallInfo&     call_info,
+                                              const std::string_view object_type,
+                                              const format::HandleId object_id,
+                                              const std::string_view command_name);
+
+    void WriteMarker(const char* name, const std::string_view marker_type, uint64_t frame_number);
+
+    /// @brief Output the boilerplate for representing a metadata block in JSON,
+    /// returning the root of the empty "args" JSON sub-tree for the caller to populate.
+    nlohmann::ordered_json& WriteMetaCommandStart(const std::string_view command_name);
+
+    /// @brief Output the boilerplate for converting a metadata block to JSON but
+    /// defer internals to the function passed in.
+    /// @tparam ToJsonFunctionType A callable which fills out the body of the object.
+    /// @todo Make this a regular function with a callback or just have callers use WriteMetaCommandStart() and
+    /// WriteBlockEnd().
```

Comment:
/// @todo Make  callers use WriteMetaCommandStart() and WriteBlockEnd() directly instead. It is no more code for them and has more straightforward code flow.